# 错误处理

### Result<T,E>

```rust
enum Result<T, E> {
    Ok(T),
    Err(E),
}
```

#### 失败就panic: unwrap 或者 expect

```rust
// unwrap 如果Ok, 返回Ok的值， 如果Err,则调用panic
// expect 出错则直接panic
// ? 出错则直接返回错误
use std::fs::File;
use std::io;
use std::io::Read;

fn read_unsename_from_file() -> Result<String, io::Error> {
    let mut s = String::new();
    
    File::open("hello.txt")?.read_to_string(&mut s)?;
}
```

#### 传播错误-？

- ？的错误使用

```rust

// ? 操作符需要一个变量来承载正确的值. ？ 用于一下形式
let v = xxx()?;
xxx()?.yyy()?;
```

使用实例：

```rust
use std::fs::File;
use std::io;
use std::io::Read;

fn read_username_from_file() -> Result<String, io::Error> {
    let mut s = String::new();
    
    File::open("hello.txt")?.read_to_string(&mut s)?;
    Ok(s)
}
```

```rust
// 骚操作
use std::fs;
use std::io;

fn read_unsername_from_file() -> Result<String, io::Error> {
    fs::read_to_string("hello.txt")
}
```

### Panic

```rust
fn main() {
    panic!("xxx");
}
```
